src/smt_util/node_traversal.h

```diff
@@ -0,0 +1,117 @@
+/*********************                                                        */
+/*! \file node_traversal.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Iterators for traversing nodes.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__SMT_UTIL__NODE_TRAVERSAL_H
+#define CVC4__SMT_UTIL__NODE_TRAVERSAL_H
+
+#include <cstddef>
+#include <iterator>
+#include <vector>
+
+#include "expr/node.h"
+
+namespace CVC4 {
+
+// Iterator for traversing a node in post-order
+// It does DAG-traversal, so indentical sub-nodes will be visited once only.
+class NodePostorderIterator
```

Comment:
How about making the iterator more general. E.g., a `NodeDfsIterator` that can do both pre-order and post-order traversal. The constructor can take a bool flag that indicates whether you want pre or post-order traversals.

Reply:
A good idea, and what I plan to do.

I thought that'd I'd start with something simple (just post-order), look for use-sites, discover which generalizations would be useful (e.g. pre-order traversal), expand the interface, and then rinse-and-repeat.

Reply:
Done in c934b96ec and 8016712f4